_profiles/drupal/resource-versioning/index.md

```diff
@@ -0,0 +1,273 @@
+    resource requested.
+  - The server **MUST NOT** include a version of a resource inappropriate for the
+    requested version.
+  - The server **MUST** respond with `404 Not Found` if an appropriate version of
+    the resource requested cannot be located. 
+    
+> Note: This means that a server should not provide "fallbacks" unless the
+> behavior is well defined by the version negotiation mechanism (see below).
+> These rules apply to individual and collection endpoints alike.
+
+## Format
+  
+The value of the `resourceVersion` parameter **MUST** be a colon-separated
+(U+003A COLON, “:”) string. The first segment of the string **SHOULD** be
+interpreted as an identifier for a _version negotiation mechanism_. A version
+negotiation mechanism defines how a server will locate an appropriate resource
+version. Subsequent segments of the string **SHOULD** be interpreted as version
+negotiation arguments for the preceding mechanism. Collectively, this query
+parameter value is known as the _version identifier_.
+
+> Note: For example, a server may support both ID-based and time-based
+> mechanisms for requesting a resource version. The former mechanism would be
+> useful for comparing versions and the latter could be useful for requesting a
+> resource as it existed at an arbitrary point in time. This profile does not
+> attempt to define every possible mechanism for versioning resources.
+
+```
+                   version-identifier
+                   _______|_________
+                  /                \
+?resourceVersion=rel:latest-version
+                  \_/ \____________/
+                   |        |
+         version-negotiator |
+                     version-argument
+```
+
+## Server Responsibilities
+
+<a id="bad-version-negotiator"></a>A server **MUST** respond with `400 Bad Request` if a version negotiator is not
+supported. In this case, an error object that includes a `type` link to
+`https://jsonapi.org/profiles/drupal/resource-versioning/#bad-version-negotiator`
+**MUST** be included in the response document.
+   
+<a id="bad-version-argument"></a>If a server cannot process the given version argument for the given negotiation
+mechanism, it **MUST** respond with a `400 Bad Request`. In this case, an error
+object that includes a `type` link to
+`https://jsonapi.org/profiles/drupal/resource-versioning/#bad-version-argument`
+**MUST** be included in the response document.
+
+If a server is able to process the version argument but an appropriate version
+cannot be located, the server **MUST** respond with a `404 Not Found`.
+
+# Links
+
+When a server processes a request with a `resourceVersion` query parameter and
+a `self` link is provided for a top-level links object, the link's `href`
+**MUST** include the `resourceVersion` query parameter with the same version
+identifier that was requested.
+
+When a server processes a request with a `resourceVersion` query parameter
```

Comment:
> which should probably usually be immutable

I think it makes sense to treat revisions as immutable.

----

I agree with all that was said in the comment above, but I want to highlight another implicit way of creating revisions that we may need to be explicit about.

Many frameworks handle revision creation as part of the store update. In Drupal when you save an entity that supports revisioning by `PATCH`ing it (no `resourceVersion` used here), then a new revision will be created for you. You cannot do much about it in most of the cases. This new revision will be a _working-copy_ or a _version_ depending on some custom business logic (usually a _published: true / false_ flag).

I guess that what I'm saying is that strict regulation of how revisions are created may become hard. However, we can put language on how to create revisions **through JSON:API**.

Where I'm going is, do we want this profile to include revision creation or it can be a separate profile?

Reply:
> [Not] addressing [PATCH/DELETE] here would seem like a major missed opportunity that might be hard to remedy once this is published.

I think I agree.

> Where I'm going is, do we want this profile to include revision creation or it can be a separate profile?

As above, I think I _would_ like this profile to define how update/delete should be handled.

<hr/>

> I think the considerations about updating bear directly on what the `self` link should be. If the self link is URL of a specific revision (as it seems to be here), then a PATCH to that link would be interpreted as updating the revision -- which should probably usually be immutable -- and creating a new revision would become a POST (or a patch to the default version's URL).

I think I may be in a minority about this, but I try to deeply appreciate the difference between a top-level `self` link and a resource object `self` link. I feel that once one groks what it means when they're different, the way JSON:API reconciles resource objects with HTTP resources + entities becomes much easier to think about.

Where I'm going with that is pretty simple: the top-level `self` link should be the URL used to mutate a resource object, not the resource object self link. Once you establish that, things become pretty simple.

```
GET /article/1?resourceVersion=rel:working-copy

{
  "data": {
    "type": "article",
    "id": 1,
    "links": {
      "self": "/article/1?resourceVersion=id:42"
    }
  },
  "links": {
    "self": "/article/1?resourceVersion=rel:working-copy"
  }
}
```

That means that a `PATCH` would update the working copy, whatever it may be.

The response would then be:

```
200 OK

{
  "data": {
    "type": "article",
    "id": 1,
    "links": {
      "self": "/article/1?resourceVersion=id:43" // <- incremented
    }
  },
  "links": {
    "self": "/article/1?resourceVersion=rel:working-copy" // <- unchanged
  }
}
```

>> which should probably usually be immutable
>
> I think it makes sense to treat revisions as immutable.

I think it makes sense too, but I don't think this profile needs to define that. I think it can work either way.

Drupal actually _does_ permit updating a revision in-place (whether that's a good idea or not is beside the point 😛 ). With the scheme above, it would be possible for a server to support both in-place editing of a revision or new revision creation via `PATCH` depending on the `self` link used (top-level vs. resource object).

If a server cannot automatically create new revisions via `PATCH` for some reason (maybe it can't/hasn't implemented the `rel` negotiator), then I think the appropriate URL for creating new revisions would be at the `version-history` URL via `POST`.

Reply:
The language about updating resources in the spec seems to jive with what I suggested:

>The URL for a resource can be obtained in the `self` link of the resource object. **Alternatively, when a `GET` request returns a single resource object as primary data, the same request URL can be used for updates**.

Reply:
I agree with @ethanresnick that the fact that this profile's narrowness (only defining "read" behavior) is its biggest weakness.

(Similarly, I think #824's biggest weakness _also_ was its narrowness, it just was different: it focused on *optimistic concurrency control*.)

On the other hand, I think solid read-only support and leaving modifications to a separate profile or a future iteration is preferable over having a spec that supports fewer use cases (which I think was true of #824 — _it did not allow a particular revision to be retrieved_). Especially if it's based on another established standard (RFC5829) and therefore likely to be more implementation-agnostic.

Reply:
I'm fascinated by @gabesullice's proposal (and the sample response bodies really help, thanks! 🙏). I'm very curious to find out what @ethanresnick thinks about that :)

Reply:
> reject a PATCH if the resource's working copy has changed from the working copy that the client is trying to apply its PATCH to

Support a conflict detection mechanism as described in initial review post wasn't discussed enough in my opinion. Something similar to AWS Amplify should be easy to achieve if already having well defined revisions. It was already discussed as part of #824 some time ago.

Maybe something like this would be everything needed:

> # Conflict detection when updating resources
>
> A server **SHOULD** include the most specific version negotiator it supports in any resource object's meta information under `revision` key.
> A client **MAY** include the revision an update is based on as under `revision` key of `meta` object on a `PATCH` request.
> A server **MUST** reject `409 Conflict` when processing a `PATCH` request to update a resource if a `revision` is provided as meta information, which is not the latest one.
> A server **MAY** reject a `PATCH` request to update a resource if server has included a revision on retrieving the resource but the `PATCH` request does not include a revision.